File: packages/firebase_common/lib/src/firestore_documents/chat_message.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'chat_message.dart';

class ReadChatMessage {
  const ReadChatMessage({
    required this.chatMessageId,
    required this.path,
    required this.senderId,
    required this.chatMessageType,
    required this.content,
    required this.imageUrls,
    required this.isDeleted,
    required this.createdAt,
    required this.updatedAt,
  });

  final String chatMessageId;

  final String path;

  final String senderId;

  final ChatMessageType chatMessageType;

  final String content;

  final List<String> imageUrls;

  final bool isDeleted;

  final DateTime? createdAt;

  final DateTime? updatedAt;

  factory ReadChatMessage._fromJson(Map<String, dynamic> json) {
    return ReadChatMessage(
      chatMessageId: json['chatMessageId'] as String,
      path: json['path'] as String,
      senderId: json['senderId'] as String,
      chatMessageType:
          _chatMessageTypeConverter.fromJson(json['chatMessageType'] as String),
      content: json['content'] as String? ?? '',
      imageUrls: (json['imageUrls'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          const <String>[],
      isDeleted: json['isDeleted'] as bool? ?? false,
      createdAt: (json['createdAt'] as Timestamp?)?.toDate(),
      updatedAt: (json['updatedAt'] as Timestamp?)?.toDate(),
    );
  }

  factory ReadChatMessage.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadChatMessage._fromJson(<String, dynamic>{
      ...data,
      'chatMessageId': ds.id,
      'path': ds.reference.path,
    });
  }
}

class CreateChatMessage {
  const CreateChatMessage({
    required this.senderId,
    required this.chatMessageType,
    required this.content,
    this.imageUrls = const <String>[],
    this.isDeleted = false,
  });

  final String senderId;
  final ChatMessageType chatMessageType;
  final String content;
  final List<String> imageUrls;
  final bool isDeleted;

  Map<String, dynamic> toJson() {
    return {
      'senderId': senderId,
      'chatMessageType': _chatMessageTypeConverter.toJson(chatMessageType),
      'content': content,
      'imageUrls': imageUrls,
      'isDeleted': isDeleted,
      'createdAt': FieldValue.serverTimestamp(),
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class UpdateChatMessage {
  const UpdateChatMessage({
    this.senderId,
    this.chatMessageType,
    this.content,
    this.imageUrls,
    this.isDeleted,
    this.createdAt,
  });

  final String? senderId;
  final ChatMessageType? chatMessageType;
  final String? content;
  final List<String>? imageUrls;
  final bool? isDeleted;
  final DateTime? createdAt;

  Map<String, dynamic> toJson() {
    return {
      if (senderId != null) 'senderId': senderId,
      if (chatMessageType != null)
        'chatMessageType': _chatMessageTypeConverter.toJson(chatMessageType!),
      if (content != null) 'content': content,
      if (imageUrls != null) 'imageUrls': imageUrls,
      if (isDeleted != null) 'isDeleted': isDeleted,
      if (createdAt != null) 'createdAt': createdAt,
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class DeleteChatMessage {}

/// Provides a reference to the chatMessages collection for reading.
CollectionReference<ReadChatMessage> readChatMessageCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('chatMessages')
        .withConverter<ReadChatMessage>(
          fromFirestore: (ds, _) => ReadChatMessage.fromDocumentSnapshot(ds),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a chatMessage document for reading.
DocumentReference<ReadChatMessage> readChatMessageDocumentReference({
  required String chatRoomId,
  required String chatMessageId,
}) =>
    readChatMessageCollectionReference(chatRoomId: chatRoomId)
        .doc(chatMessageId);

/// Provides a reference to the chatMessages collection for creating.
CollectionReference<CreateChatMessage> createChatMessageCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('chatMessages')
        .withConverter<CreateChatMessage>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a chatMessage document for creating.
DocumentReference<CreateChatMessage> createChatMessageDocumentReference({
  required String chatRoomId,
  required String chatMessageId,
}) =>
    createChatMessageCollectionReference(chatRoomId: chatRoomId)
        .doc(chatMessageId);

/// Provides a reference to the chatMessages collection for updating.
CollectionReference<UpdateChatMessage> updateChatMessageCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('chatMessages')
        .withConverter<UpdateChatMessage>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a chatMessage document for updating.
DocumentReference<UpdateChatMessage> updateChatMessageDocumentReference({
  required String chatRoomId,
  required String chatMessageId,
}) =>
    updateChatMessageCollectionReference(chatRoomId: chatRoomId)
        .doc(chatMessageId);

/// Provides a reference to the chatMessages collection for deleting.
CollectionReference<DeleteChatMessage> deleteChatMessageCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('chatMessages')
        .withConverter<DeleteChatMessage>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a chatMessage document for deleting.
DocumentReference<DeleteChatMessage> deleteChatMessageDocumentReference({
  required String chatRoomId,
  required String chatMessageId,
}) =>
    deleteChatMessageCollectionReference(chatRoomId: chatRoomId)
        .doc(chatMessageId);

/// Manages queries against the chatMessages collection.
class ChatMessageQuery {
  /// Fetches [ReadChatMessage] documents.
  Future<List<ReadChatMessage>> fetchDocuments({
    required String chatRoomId,
    GetOptions? options,
    Query<ReadChatMessage>? Function(Query<ReadChatMessage> query)?
        queryBuilder,
    int Function(ReadChatMessage lhs, ReadChatMessage rhs)? compare,
  }) async {
    Query<ReadChatMessage> query =
        readChatMessageCollectionReference(chatRoomId: chatRoomId);
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [ChatMessage] documents.
  Stream<List<ReadChatMessage>> subscribeDocuments({
    required String chatRoomId,
    Query<ReadChatMessage>? Function(Query<ReadChatMessage> query)?
        queryBuilder,
    int Function(ReadChatMessage lhs, ReadChatMessage rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadChatMessage> query =
        readChatMessageCollectionReference(chatRoomId: chatRoomId);
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadChatMessage] document.
  Future<ReadChatMessage?> fetchDocument({
    required String chatRoomId,
    required String chatMessageId,
    GetOptions? options,
  }) async {
    final ds = await readChatMessageDocumentReference(
      chatRoomId: chatRoomId,
      chatMessageId: chatMessageId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [ChatMessage] document.
  Stream<ReadChatMessage?> subscribeDocument({
    required String chatRoomId,
    required String chatMessageId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readChatMessageDocumentReference(
      chatRoomId: chatRoomId,
      chatMessageId: chatMessageId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [ChatMessage] document.
  Future<DocumentReference<CreateChatMessage>> add({
    required String chatRoomId,
    required CreateChatMessage createChatMessage,
  }) =>
      createChatMessageCollectionReference(chatRoomId: chatRoomId)
          .add(createChatMessage);

  /// Sets a [ChatMessage] document.
  Future<void> set({
    required String chatRoomId,
    required String chatMessageId,
    required CreateChatMessage createChatMessage,
    SetOptions? options,
  }) =>
      createChatMessageDocumentReference(
        chatRoomId: chatRoomId,
        chatMessageId: chatMessageId,
      ).set(createChatMessage, options);

  /// Updates a specific [ChatMessage] document.
  Future<void> update({
    required String chatRoomId,
    required String chatMessageId,
    required UpdateChatMessage updateChatMessage,
  }) =>
      updateChatMessageDocumentReference(
        chatRoomId: chatRoomId,
        chatMessageId: chatMessageId,
      ).update(updateChatMessage.toJson());

  /// Deletes a specific [ChatMessage] document.
  Future<void> delete({
    required String chatRoomId,
    required String chatMessageId,
  }) =>
      deleteChatMessageDocumentReference(
        chatRoomId: chatRoomId,
        chatMessageId: chatMessageId,
      ).delete();
}

File: functions/src/firestore-documents/chatMessage.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    CollectionReference,
    DocumentData,
    DocumentReference,
    DocumentSnapshot,
    FieldValue,
    Query,
    QueryDocumentSnapshot,
    QuerySnapshot,
    SetOptions,
    Timestamp,
    WriteResult
} from 'firebase-admin/firestore'

export type ChatMessageType = `worker` | `host` | `system`

export class ReadChatMessage {
    constructor({
        chatMessageId,
        path,
        senderId,
        chatMessageType,
        content,
        imageUrls,
        isDeleted,
        createdAt,
        updatedAt
    }: {
        chatMessageId: string
        path: string
        senderId: string
        chatMessageType: ChatMessageType
        content: string
        imageUrls: string[]
        isDeleted: boolean
        createdAt?: Date
        updatedAt?: Date
    }) {
        this.chatMessageId = chatMessageId
        this.path = path
        this.senderId = senderId
        this.chatMessageType = chatMessageType
        this.content = content
        this.imageUrls = imageUrls
        this.isDeleted = isDeleted
        this.createdAt = createdAt
        this.updatedAt = updatedAt
    }

    readonly chatMessageId: string

    readonly path: string

    readonly senderId: string

    readonly chatMessageType: ChatMessageType

    readonly content: string

    readonly imageUrls: string[]

    readonly isDeleted: boolean

    readonly createdAt?: Date

    readonly updatedAt?: Date

    private static chatMessageTypeConverterFromJson(
        chatMessageType: string
    ): ChatMessageType {
        return chatMessageType as ChatMessageType
    }

    private static fromJson(json: Record<string, unknown>): ReadChatMessage {
        return new ReadChatMessage({
            chatMessageId: json[`chatMessageId`] as string,
            path: json[`path`] as string,
            senderId: json[`senderId`] as string,
            chatMessageType: ReadChatMessage.chatMessageTypeConverterFromJson(
                json[`chatMessageType`] as string
            ),
            content: (json[`content`] as string | undefined) ?? ``,
            imageUrls:
                (json[`imageUrls`] as unknown[] | undefined)?.map(
                    (e) => e as string
                ) ?? [],
            isDeleted: (json[`isDeleted`] as boolean | undefined) ?? false,
            createdAt: (json[`createdAt`] as Timestamp | undefined)?.toDate(),
            updatedAt: (json[`updatedAt`] as Timestamp | undefined)?.toDate()
        })
    }

    static fromDocumentSnapshot(ds: DocumentSnapshot): ReadChatMessage {
        const data = ds.data()!
        const cleanedData: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(data)) {
            cleanedData[key] = value === null ? undefined : value
        }
        return ReadChatMessage.fromJson({
            ...cleanedData,
            chatMessageId: ds.id,
            path: ds.ref.path
        })
    }
}

export class CreateChatMessage {
    constructor({
        senderId,
        chatMessageType,
        content,
        imageUrls,
        isDeleted
    }: {
        senderId: string
        chatMessageType: ChatMessageType
        content: string
        imageUrls: string[]
        isDeleted: boolean
    }) {
        this.senderId = senderId
        this.chatMessageType = chatMessageType
        this.content = content
        this.imageUrls = imageUrls
        this.isDeleted = isDeleted
    }

    readonly senderId: string

    readonly chatMessageType: ChatMessageType

    readonly content: string

    readonly imageUrls: string[]

    readonly isDeleted: boolean

    private static chatMessageTypeConverterToJson(
        chatMessageType: ChatMessageType
    ): string {
        return chatMessageType as string
    }

    toJson(): Record<string, unknown> {
        return {
            senderId: this.senderId,
            chatMessageType: CreateChatMessage.chatMessageTypeConverterToJson(
                this.chatMessageType
            ),
            content: this.content,
            imageUrls: this.imageUrls,
            isDeleted: this.isDeleted,
            createdAt: FieldValue.serverTimestamp(),
            updatedAt: FieldValue.serverTimestamp()
        }
    }
}

export class UpdateChatMessage {
    constructor({
        senderId,
        chatMessageType,
        content,
        imageUrls,
        isDeleted,
        createdAt
    }: {
        senderId?: string
        chatMessageType?: ChatMessageType
        content?: string
        imageUrls?: string[]
        isDeleted?: boolean
        createdAt?: Date
    }) {
        this.senderId = senderId
        this.chatMessageType = chatMessageType
        this.content = content
        this.imageUrls = imageUrls
        this.isDeleted = isDeleted
        this.createdAt = createdAt
    }

    readonly senderId?: string

    readonly chatMessageType?: ChatMessageType

    readonly content?: string

    readonly imageUrls?: string[]

    readonly isDeleted?: boolean

    readonly createdAt?: Date

    private static chatMessageTypeConverterToJson(
        chatMessageType: ChatMessageType
    ): string {
        return chatMessageType as string
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toJson(): Record<string, any> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const json: Record<string, any> = {}
        if (this.senderId != undefined) {
            json[`senderId`] = this.senderId
        }
        if (this.chatMessageType != undefined) {
            json[`chatMessageType`] =
                UpdateChatMessage.chatMessageTypeConverterToJson(
                    this.chatMessageType
                )
        }
        if (this.content != undefined) {
            json[`content`] = this.content
        }
        if (this.imageUrls != undefined) {
            json[`imageUrls`] = this.imageUrls
        }
        if (this.isDeleted != undefined) {
            json[`isDeleted`] = this.isDeleted
        }
        if (this.createdAt != undefined) {
            json[`createdAt`] = this.createdAt
        }
        json[`updatedAt`] = FieldValue.serverTimestamp()
        return json
    }
}

export class DeleteChatMessage {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the chatMessages collection for reading.
 */
export const readChatMessageCollectionReference = ({
    chatRoomId
}: {
    chatRoomId: string
}): CollectionReference<ReadChatMessage> => {
    return db
        .collection(`chatRooms`)
        .doc(chatRoomId)
        .collection(`chatMessages`)
        .withConverter<ReadChatMessage>({
            fromFirestore: (ds: DocumentSnapshot): ReadChatMessage => {
                return ReadChatMessage.fromDocumentSnapshot(ds)
            },
            toFirestore: () => {
                throw new Error(
                    `toFirestore is not implemented for ReadChatMessage`
                )
            }
        })
}

/**
 * Provides a reference to a chatMessage document for reading.
 * @param chatMessageId - The ID of the chatMessage document to read.
 */
export const readChatMessageDocumentReference = ({
    chatRoomId,
    chatMessageId
}: {
    chatRoomId: string
    chatMessageId: string
}): DocumentReference<ReadChatMessage> =>
    readChatMessageCollectionReference({
        chatRoomId
    }).doc(chatMessageId)

/**
 * Provides a reference to the chatMessages collection for creating.
 */
export const createChatMessageCollectionReference = ({
    chatRoomId
}: {
    chatRoomId: string
}): CollectionReference<CreateChatMessage> => {
    return db
        .collection(`chatRooms`)
        .doc(chatRoomId)
        .collection(`chatMessages`)
        .withConverter<CreateChatMessage>({
            fromFirestore: () => {
                throw new Error(
                    `fromFirestore is not implemented for CreateChatMessage`
                )
            },
            toFirestore: (obj: CreateChatMessage): DocumentData => {
                return obj.toJson()
            }
        })
}

/**
 * Provides a reference to a chatMessage document for creating.
 * @param chatMessageId - The ID of the chatMessage document to read.
 */
export const createChatMessageDocumentReference = ({
    chatRoomId,
    chatMessageId
}: {
    chatRoomId: string
    chatMessageId: string
}): DocumentReference<CreateChatMessage> =>
    createChatMessageCollectionReference({
        chatRoomId
    }).doc(chatMessageId)

/**
 * Provides a reference to the chatMessages collection for updating.
 */
export const updateChatMessageCollectionReference = ({
    chatRoomId
}: {
    chatRoomId: string
}): CollectionReference<UpdateChatMessage> => {
    return db
        .collection(`chatRooms`)
        .doc(chatRoomId)
        .collection(`chatMessages`)
        .withConverter<UpdateChatMessage>({
            fromFirestore: () => {
                throw new Error(
                    `fromFirestore is not implemented for UpdateChatMessage`
                )
            },
            toFirestore: (obj: UpdateChatMessage): DocumentData => {
                return obj.toJson()
            }
        })
}

/**
 * Provides a reference to a chatMessage document for updating.
 * @param chatMessageId - The ID of the chatMessage document to read.
 */
export const updateChatMessageDocumentReference = ({
    chatRoomId,
    chatMessageId
}: {
    chatRoomId: string
    chatMessageId: string
}): DocumentReference<UpdateChatMessage> =>
    updateChatMessageCollectionReference({
        chatRoomId
    }).doc(chatMessageId)

/**
 * Provides a reference to the chatMessages collection for deleting.
 */
export const deleteChatMessageCollectionReference = ({
    chatRoomId
}: {
    chatRoomId: string
}): CollectionReference<DeleteChatMessage> => {
    return db
        .collection(`chatRooms`)
        .doc(chatRoomId)
        .collection(`chatMessages`)
        .withConverter<DeleteChatMessage>({
            fromFirestore: () => {
                throw new Error(
                    `fromFirestore is not implemented for DeleteChatMessage`
                )
            },
            toFirestore: (): DocumentData => {
                throw new Error(
                    `toFirestore is not implemented for DeleteChatMessage`
                )
            }
        })
}

/**
 * Provides a reference to a chatMessage document for deleting.
 * @param chatMessageId - The ID of the chatMessage document to read.
 */
export const deleteChatMessageDocumentReference = ({
    chatRoomId,
    chatMessageId
}: {
    chatRoomId: string
    chatMessageId: string
}): DocumentReference<DeleteChatMessage> =>
    deleteChatMessageCollectionReference({
        chatRoomId
    }).doc(chatMessageId)

/**
 * Manages queries against the chatMessages collection.
 */
export class ChatMessageQuery {
    /**
     * Fetches chatMessage documents.
     * @param queryBuilder - Function to modify the query.
     * @param compare - Function to sort the results.
     */
    async fetchDocuments({
        chatRoomId,
        queryBuilder,
        compare
    }: {
        chatRoomId: string
        queryBuilder?: (query: Query<ReadChatMessage>) => Query<ReadChatMessage>
        compare?: (lhs: ReadChatMessage, rhs: ReadChatMessage) => number
    }): Promise<ReadChatMessage[]> {
        let query: Query<ReadChatMessage> = readChatMessageCollectionReference({
            chatRoomId
        })
        if (queryBuilder != undefined) {
            query = queryBuilder(query)
        }
        const qs: QuerySnapshot<ReadChatMessage> = await query.get()
        let result = qs.docs.map(
            (qds: QueryDocumentSnapshot<ReadChatMessage>) => qds.data()
        )
        if (compare != undefined) {
            result = result.sort(compare)
        }
        return result
    }

    /**
     * Fetches a specific chatMessage document.
     * @param chatMessageId - The ID of the chatMessage document to fetch.
     */
    async fetchDocument({
        chatRoomId,
        chatMessageId
    }: {
        chatRoomId: string
        chatMessageId: string
    }): Promise<ReadChatMessage | undefined> {
        const ds = await readChatMessageDocumentReference({
            chatRoomId,
            chatMessageId
        }).get()
        return ds.data()
    }

    /**
     * Adds a chatMessage document.
     * @param createChatMessage - The chatMessage details to add.
     */
    async add({
        chatRoomId,
        createChatMessage
    }: {
        chatRoomId: string
        createChatMessage: CreateChatMessage
    }): Promise<DocumentReference<CreateChatMessage>> {
        return createChatMessageCollectionReference({ chatRoomId }).add(
            createChatMessage
        )
    }

    /**
     * Sets a chatMessage document.
     * @param chatMessageId - The ID of the chatMessage document to set.
     * @param createChatMessage - The chatMessage details to set.
     * @param options - Options for the set operation.
     */
    async set({
        chatRoomId,
        chatMessageId,
        createChatMessage,
        options
    }: {
        chatRoomId: string
        chatMessageId: string
        createChatMessage: CreateChatMessage
        options?: SetOptions
    }): Promise<WriteResult> {
        if (options == undefined) {
            return createChatMessageDocumentReference({
                chatRoomId,
                chatMessageId
            }).set(createChatMessage)
        } else {
            return createChatMessageDocumentReference({
                chatRoomId,
                chatMessageId
            }).set(createChatMessage, options)
        }
    }

    /**
     * Updates a specific chatMessage document.
     * @param chatMessageId - The ID of the chatMessage document to update.
     * @param updateChatMessage - The details for updating the chatMessage.
     */
    async update({
        chatRoomId,
        chatMessageId,
        updateChatMessage
    }: {
        chatRoomId: string
        chatMessageId: string
        updateChatMessage: UpdateChatMessage
    }): Promise<WriteResult> {
        return updateChatMessageDocumentReference({
            chatRoomId,
            chatMessageId
        }).update(updateChatMessage.toJson())
    }

    /**
     * Deletes a specific chatMessage document.
     * @param chatMessageId - The ID of the chatMessage document to delete.
     */
    async delete({
        chatRoomId,
        chatMessageId
    }: {
        chatRoomId: string
        chatMessageId: string
    }): Promise<WriteResult> {
        return deleteChatMessageDocumentReference({
            chatRoomId,
            chatMessageId
        }).delete()
    }
}
